docs/manpages/mvxattr.1
'\" t
.\"     Title: mvxattr
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 12/15/2022
.\"    Manual: User Commands
.\"    Source: Samba 4.15.13
.\"  Language: English
.\"
.TH "MVXATTR" "1" "12/15/2022" "Samba 4\&.15\&.13" "User Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mvxattr \- Recursively rename extended attributes
.SH "SYNOPSIS"
.HP \w'\ 'u
mvxattr {\-s\ STRING,\ \-\-from=STRING} {\-d\ STRING,\ \-\-to=STRING} [\-l,\ \-\-follow\-symlinks] [\-p,\ \-\-print] [\-v,\ \-\-verbose] [\-f,\ \-\-force] {PATH\ [PATH\ \&.\&.\&.]}
.SH "DESCRIPTION"
.PP
This tool is part of the
\fBsamba\fR(1)
suite\&.
.PP
mvxattr is a simple utility to recursively rename extended attributes\&.
.PP
By default all symlinks are ignored, use
\fB\-l\fR
to follow them\&.
.SH "OPTIONS"
.PP
\-s STRING, \-\-from=STRING
.RS 4
Source xattr name
.RE
.PP
\-d STRING, \-\-to=STRING
.RS 4
Destination xattr name
.RE
.PP
\-l, \-\-follow\-symlinks
.RS 4
Follow symlinks, the default is to ignore them\&.
.RE
.PP
\-p, \-\-print
.RS 4
Print files where the xattr got renamed\&.
.RE
.PP
\-v, \-\-verbose
.RS 4
Print files as they are checked\&.
.RE
.PP
\-f, \-\-force
.RS 4
Force overwriting of destination xattr\&.
.RE
.SH "VERSION"
.PP
This man page is part of version 4\&.15\&.13 of the Samba suite\&.
.SH "AUTHOR"
.PP
The original Samba software and related utilities were created by Andrew Tridgell\&. Samba is now developed by the Samba Team as an Open Source project similar to the way the Linux kernel is developed\&.
.PP
The mvxattr manpage was written by Ralph Boehme\&.


debian/NEWS
samba (2:4.4.1+dfsg-1) experimental; urgency=medium

    This Samba security addresses both Denial of Service and Man in
    the Middle vulnerabilities.

    Both of these changes implement new smb.conf options and a number
    of stricter behaviours to prevent Man in the Middle attacks on our
    network services, as a client and as a server.

    Between these changes, compatibility with a large number of older
    software versions has been lost in the default configuration.

    See the release notes in WHATNEW.txt for more information.


    Here are some additional hints how to work around the new stricter default behaviors:

    * As an AD DC server, only Windows 2000 and Samba 3.6 and above as
      a domain member are supported out of the box. Other smb file
      servers as domain members are also fine out of the box.

    * As an AD DC server, with default setting of "ldap server require
      strong auth", LDAP clients connecting over ldaps:// or START_TLS
      will be allowed to perform simple LDAP bind only.

      The preferred configuration for LDAP clients is to use SASL
      GSSAPI directly over ldap:// without using ldaps:// or
      START_TLS.

      To use LDAP with START_TLS and SASL GSSAPI (either Kerberos or
      NTLMSSP) sign/seal protection must be used by the client and
      server should be configured with "ldap server require strong
      auth = allow_sasl_over_tls".

      Consult OpenLDAP documentation how to set sign/seal protection
      in ldap.conf.

      For SSSD client configured with "id_provider = ad" or
      "id_provider = ldap" with "auth_provider = krb5", see
      sssd-ldap(5) manual for details on TLS session handling.

    * As a File Server, compatibility with the Linux Kernel cifs
      client depends on which configuration options are selected, please
      use "sec=krb5(i)" or "sec=ntlmssp(i)", not "sec=ntlmv2".

    * As a file or printer client and as a domain member, out of the
      box compatibility with Samba less than 4.0 and other SMB/CIFS
      servers, depends on support for SMB signing or SMB2 on the
      server, which is often disabled or absent. You may need to
      adjust the "client ipc signing" to "no" in these cases.

    * In case of an upgrade from versions before 4.2.0, you might run
      into problems as a domain member. The out of the box compatibility
      with Samba 3.x domain controllers requires NETLOGON features only
      available in Samba 3.2 and above.

    However, all of these can be worked around by setting smb.conf
    options in Samba, see WHATSNEW.txt the 4.2.0 release notes at
    https://www.samba.org/samba/history/samba-4.2.0.html and the Samba
    wiki for details, workarounds and suggested security-improving
    changes to these and other software packages.


    Suggested further improvements after patching:

    It is recommended that administrators set these additional options,
    if compatible with their network environment:

        server signing = mandatory
        ntlm auth = no

    Without "server signing = mandatory", Man in the Middle attacks
    are still possible against our file server and
    classic/NT4-like/Samba3 Domain controller. (It is now enforced on
    Samba's AD DC.) Note that this has heavy impact on the file server
    performance, so you need to decide between performance and
    security. These Man in the Middle attacks for smb file servers are
    well known for decades.

    Without "ntlm auth = no", there may still be clients not using
    NTLMv2, and these observed passwords may be brute-forced easily using
    cloud-computing resources or rainbow tables.

 -- Andrew Bartlett <abartlet+debian@catalyst.net.nz>  Tue, 12 Apr 2016 16:18:57 +1200

samba (2:4.0.10+dfsg-3) unstable; urgency=low

    The SWAT package is no longer available.

    Upstream support for SWAT (Samba Web Administration Tool) was removed in
    samba 4.1.0. As a result, swat is no longer shipped in the Debian Samba
    packages. Unfortunately, there is currently no replacement.

    Details why SWAT has been removed upstream can be found on the
    samba-technical mailing list:

    https://lists.samba.org/archive/samba-technical/2013-February/090572.html

 -- Ivo De Decker <ivo.dedecker@ugent.be>  Tue, 22 Oct 2013 07:52:54 +0200

samba (2:3.6.5-2) unstable; urgency=low

    NSS modules have been split out from libpam-winbind to
    libnss-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libnss-winbind package after upgrading
    from former versions of winbind (for instance from squeeze) or
    from former versions of libpam-winbind.

 -- Christian Perrier <bubulle@debian.org>  Mon, 07 May 2012 22:16:32 +0200

samba (2:3.5.11~dfsg-3) unstable; urgency=low

    PAM modules and NSS modules have been split out from the winbind
    package into libpam-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libpam-winbind package after upgrading
    from former versions of winbind (for instance from squeeze)

 -- Steve Langasek <vorlon@debian.org>  Fri, 21 Oct 2011 20:00:13 +0000

samba (2:3.4.0-1) unstable; urgency=low

    Default passdb backend changed in samba 3.4.0 and above
  
    Beginning with samba 3.4.0, the default setting for "passdb
    backend" changed from "smbpasswd" to "tdbsam".
    
    If your smb.conf file does not have an explicit mention of
    "passdb backend" when upgrading from pre-3.4.0 versions of
    samba, it is likely that users will no longer be able to
    authenticate.
    
    As a consequence of all this, if you're upgrading from lenny
    and have no setting of "passdb backend" in smb.conf, you MUST
    add "passdb backend = smbpasswd" in order to keep your samba
    server's behaviour.
    
    As Debian packages of samba explicitly set "passdb backend = tdbsam"
    by default since etch, very few users should need to modify their

source3/utils/mvxattr.c
/*
   Unix SMB/CIFS implementation.
   xattr renaming
   Copyright (C) Ralph Boehme 2017

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "includes.h"
#include "system/filesys.h"
#include <popt.h>
#include <ftw.h>

#ifdef __OS2__
ssize_t getxattr(const char *path, const char *name,
                 void *value, size_t size);
int removexattr(const char *path, const char *name);
int setxattr(const char *path, const char *name,
              const void *value, size_t size, int flags);
#endif
static struct rename_xattr_state {
	int follow_symlink;
	int print;
	int force;
	int verbose;
	char *xattr_from;
	char *xattr_to;
} state;

static int rename_xattr(const char *path,
			const struct stat *sb,
			int typeflag,
			struct FTW *ftwbuf)
{
	ssize_t len;
	int ret;

	if (typeflag == FTW_SL) {
		d_printf("Ignoring symlink %s\n", path);
		return 0;
	}

	if (state.verbose) {
		d_printf("%s\n", path);
	}

	len = getxattr(path, state.xattr_from, NULL, 0);
	if (len < 0) {
		if (errno == ENOATTR) {
			return 0;
		}
		d_printf("getxattr [%s] failed [%s]\n",
			 path, strerror(errno));
		return -1;
	}

	{
		uint8_t buf[len];

		len = getxattr(path, state.xattr_from, &buf[0], len);
		if (len == -1) {
			d_printf("getxattr [%s] failed [%s]\n",
				 path, strerror(errno));
			return -1;
		}

		ret = setxattr(path, state.xattr_to, &buf[0], len, XATTR_CREATE);
		if (ret != 0) {
			if (errno != EEXIST) {
				d_printf("setxattr [%s] failed [%s]\n",
					 path, strerror(errno));
				return -1;
			}
			if (!state.force) {
				d_printf("destination [%s:%s] exists, use -f to force\n",
					 path, state.xattr_to);
				return -1;
			}
			ret = setxattr(path, state.xattr_to, &buf[0], len, XATTR_REPLACE);
			if (ret != 0) {
				d_printf("setxattr [%s:%s] failed [%s]\n",
					 path, state.xattr_to, strerror(errno));
				return -1;
			}
		}

		ret = removexattr(path, state.xattr_from);
		if (ret != 0) {
			d_printf("removexattr [%s:%s] failed [%s]\n",
				 path, state.xattr_from, strerror(errno));
			return -1;
		}

		if (state.print) {
			d_printf("Renamed %s to %s on %s\n",
				 state.xattr_from, state.xattr_to, path);
		}
	}

	return 0;
}

int main(int argc, const char *argv[])
{
	int c;
	const char *path = NULL;
	poptContext pc = NULL;
	struct poptOption long_options[] = {
		POPT_AUTOHELP
		{
			.longName   = "from",
			.shortName  = 's',
			.argInfo    = POPT_ARG_STRING,
			.arg        = &state.xattr_from,
			.val        = 's',
			.descrip    = "xattr source name",
		},
		{
			.longName   = "to",
			.shortName  = 'd',
			.argInfo    = POPT_ARG_STRING,
			.arg        = &state.xattr_to,
			.val        = 'd',
			.descrip    = "xattr destination name",
		},
		{
			.longName   = "follow-symlinks",
			.shortName  = 'l',
			.argInfo    = POPT_ARG_NONE,
			.arg        = &state.follow_symlink,
			.val        = 'l',
			.descrip    = "follow symlinks, the default is to "
				      "ignore them",
		},
		{
			.longName   = "print",
			.shortName  = 'p',
			.argInfo    = POPT_ARG_NONE,
			.arg        = &state.print,
			.val        = 'p',
			.descrip    = "print files where the xattr got "
				      "renamed",
		},
		{
			.longName   = "verbose",
			.shortName  = 'v',
			.argInfo    = POPT_ARG_NONE,
			.arg        = &state.verbose,
			.val        = 'v',
			.descrip    = "print files as they are checked",
		},
		{
			.longName   = "force",
			.shortName  = 'f',
			.argInfo    = POPT_ARG_NONE,
			.arg        = &state.force,
			.val        = 'f',
			.descrip    = "force overwriting of destination xattr",
		},
		POPT_TABLEEND
	};
	TALLOC_CTX *frame = talloc_stackframe();
	const char *s = NULL;
	int ret = 0;

	if (getuid() != 0) {
		d_printf("%s only works as root!\n", argv[0]);
		ret = 1;
		goto done;
	}

	pc = poptGetContext(NULL, argc, argv, long_options, 0);
	poptSetOtherOptionHelp(pc, "-s STRING -d STRING PATH [PATH ...]");

	while ((c = poptGetNextOpt(pc)) != -1) {
		switch (c) {
		case 's':
			s = poptGetOptArg(pc);
			state.xattr_from = talloc_strdup(frame, s);
			if (state.xattr_from == NULL) {
				ret = 1;
				goto done;
			}
			break;
		case 'd':
			s = poptGetOptArg(pc);
			state.xattr_to = talloc_strdup(frame, s);
			if (state.xattr_to == NULL) {
				ret = 1;
				goto done;
			}
			break;
		case POPT_ERROR_BADOPT:
			fprintf(stderr, "\nInvalid option %s: %s\n\n",
				poptBadOption(pc, 0), poptStrerror(c));
			poptPrintUsage(pc, stderr, 0);
			exit(1);
		}
	}

	if (state.xattr_from == NULL || state.xattr_to == NULL) {
		poptPrintUsage(pc, stderr, 0);
		ret = 1;
		goto done;
	}

	if (poptPeekArg(pc) == NULL) {
		poptPrintUsage(pc, stderr, 0);
		ret = 1;
		goto done;
	}

	while ((path = poptGetArg(pc)) != NULL) {
		ret = nftw(path, rename_xattr, 256,
			   state.follow_symlink ? 0 : FTW_PHYS);
	}

done:
	poptFreeContext(pc);

	TALLOC_FREE(frame);
	return ret;
}
